action.yml

```diff
@@ -77,10 +77,149 @@ runs:
           }
         }
 
+        function Invoke-Download {
+          Param
+          (
+            [Parameter(Mandatory)]
+            [String] $URL,
+            [Alias("Destination")]
+            [String] $Path,
+            [Int] $Retries = 20
+          )
+
+          if (-not $Path) {
+            $InvalidCharacters = [IO.Path]::GetInvalidFileNameChars() -Join ''
+            $RE = "[{0}]" -f [RegEx]::Escape($InvalidCharacters)
+            $FileName = [IO.Path]::GetFileName($URL) -Replace $RE
+
+            if ([String]::IsNullOrEmpty($FileName)) {
+              $FileName = [System.IO.Path]::GetRandomFileName()
+            }
+            $Path = Join-Path -Path "${env:Temp}" -ChildPath $FileName
+          }
+
+          Write-Host "Downloading package from $URL to $Path..."
+
+          $Interval = 30
+          $StartTime = Get-Date
+          do {
+            try {
+              $AttemptStartTime = Get-Date
+              (New-Object System.Net.WebClient).DownloadFile($URL, $Path)
+              $AttemptDuration = [math]::Round(($(Get-Date) - $AttemptStartTime).TotalSeconds, 2)
+              Write-Host "Package downloaded in $AttemptDuration seconds"
+              break
+            } catch {
+              $AttemptDuration = [math]::Round(($(Get-Date) - $AttemptStartTime).TotalSeconds, 2)
+              Write-Warning "Package download failed after $AttemptDuration seconds"
+              Write-Warning $_.Exception.Message
+            }
+
+            if ($Retries -eq 1) {
+              $Duration = [math]::Round(($(Get-Date) - $StartTime).TotalSeconds, 2)
+              throw "Package download failed after $Duration seconds"
+            }
+
+            Write-Warning "Waiting $Interval seconds before retrying (retries remaining: $Retries)..."
+            Start-Sleep -Seconds $Interval
+          } while (--$Retries -gt 0)
+
+          return $Path
+        }
+
+        function Verify-Checksum {
+          Param
+          (
+            [Parameter(Mandatory = $true)]
+            [String]$Actual,
+            [Parameter(Mandatory = $true)]
+            [String]$Expected
+          )
+
+          Write-Verbose "Performing Checksum Verification"
+          if ($Actual -eq $Expected) {
+            Write-Verbose "Checksum verification passed"
+          } else {
+            throw "Checksum verification failure (Actual: $Actual vs Expected: $Expected)"
+          }
+        }
+
+        function Invoke-Installation {
+          Param
+          (
+            [Parameter(Mandatory, ParameterSetName = "URL")]
+            [String] $URL,
+            [Parameter(Mandatory, ParameterSetName = "LocalPath")]
+            [String] $LocalPath,
+            [ValidateSet("MSI", "EXE")]
+            [String] $Type,
+            [String[]] $InstallArgs,
```

Comment:
Do you need this to be a param?  It gets overwritten without being used.

Reply:
Yes, its a tri-state thing.  If not specified, we default to `/i /qn /norestart`.  If specified, they are passed to the installer if it is a `.exe`.  If specified, they are taken verbatim.